DOC Add secure persistence docs

BenjaminBossan opened the pull request:
Fixes #139

merveenoyan commented on README.rst:
```diff
@@ -45,6 +45,10 @@ download or load the model.
   and how it should be used. The model card can then be stored as the
   ``README.md`` file on the Hugging Face Hub, with pre-populated metadata to
   help Hub understand the model.
+- ``skops.io``: Secure persistence of sklearn estimators and more without using
```
```suggestion
- ``skops.io``: Secure persistence of sklearn estimators and more, without using
```

merveenoyan commented on docs/persistence.rst:
```diff
@@ -0,0 +1,80 @@
+particular, this means that if you have custom functions (say, a custom function
+to be used with :class:`sklearn.preprocessing.FunctionTransformer`), it will not
+work. However, most ``numpy`` and ``scipy`` functions should work. Therefore,
+you can actually save ``numpy.sqrt`` etc.
```
```suggestion
you can actually save built-in functions like``numpy.sqrt``.
```

merveenoyan commented on docs/persistence.rst:
```diff
@@ -0,0 +1,80 @@
+-----
+
+Currently, it is still possible to run insecure code when using skops
+persistence. E.g., it's possible to load a save file that evaluates arbitrary
```
```suggestion
persistence, e.g., it's possible to load a save file that evaluates arbitrary
```

merveenoyan commented on skops/io/_persist.py:
```diff
@@ -28,6 +28,29 @@
 def save(obj, file):
+    """Save an object using the skops persistence format.
+
+    Skops aims at providing a secure persistence feature that does not rely on
```
I think it makes sense to put warnings only to load rather than save, but feel free to ignore if you think it doesn't make sense😅

BenjaminBossan replied:
I know I went overboard with warnings, but I really wanted to make sure users don't miss the fact that we're still early in development and know that it's not secure yet. Hopefully, we can fill these gaps soon and then remove the warnings. Until then, I'd rather leave them in.

merveenoyan reviewed (approved):
Left minor nits, overall it looks good but I previously read the code so I'd like @stevhliu to take a look as well!

stevhliu commented on docs/persistence.rst:
```diff
@@ -0,0 +1,80 @@
+Usage
+-----
+
+Using :func:`skops.io.save` and :func:`skops.io.load` is quite simple. Below is
```
I'd personally avoid using a "value" word like `simple` or `easy` since it is relative to a user's experience and can make the tone not as friendly for beginners. I think the code snippet below already does a great job of showing how simple it is without having us tell the user it is simple :)

stevhliu commented on docs/persistence.rst:
```diff
@@ -0,0 +1,80 @@
+At the moment, we support the vast majority of sklearn estimators. This includes
+complex use cases such as :class:`sklearn.pipeline.Pipeline`,
+:class:`sklearn.model_selection.GridSearchCV`, classes using Cython code, such
+as :class:`sklearn.tree.DecisionTreeClassifier`, etc. If you discover an sklearn
```
```suggestion
as :class:`sklearn.tree.DecisionTreeClassifier`, and more. If you discover an sklearn
```

stevhliu commented on docs/persistence.rst:
```diff
@@ -0,0 +1,80 @@
+estimator that does not work, please open an issue on the skops `GitHub page
+<https://github.com/skops-dev/skops/issues>`_ and let us know.
+
+In contrast to pickle, skops cannot persist arbitrary Python code. In
+particular, this means that if you have custom functions (say, a custom function
+to be used with :class:`sklearn.preprocessing.FunctionTransformer`), it will not
+work. However, most ``numpy`` and ``scipy`` functions should work. Therefore,
+you can actually save built-in functions like``numpy.sqrt``.
```
```suggestion
In contrast to pickle, skops cannot persist arbitrary Python code. This means 
if you have custom functions (say, a custom function to be used 
with :class:`sklearn.preprocessing.FunctionTransformer`), it will not
work. However, most ``numpy`` and ``scipy`` functions should work. Therefore,
you can actually save built-in functions like``numpy.sqrt``.
```

stevhliu commented on docs/persistence.rst:
```diff
@@ -0,0 +1,80 @@
+-----
+
+Currently, it is still possible to run insecure code when using skops
+persistence, e.g., it's possible to load a save file that evaluates arbitrary
```
```suggestion
persistence. For example, it's possible to load a save file that evaluates arbitrary
```

stevhliu commented on docs/persistence.rst:
```diff
@@ -0,0 +1,80 @@
+
+Currently, it is still possible to run insecure code when using skops
+persistence, e.g., it's possible to load a save file that evaluates arbitrary
+code using :func:`eval`. We do, however, have concrete plans on how to mitigate
+this. Please stay updated.
```
```suggestion
code using :func:`eval`. However, we have concrete plans on how to mitigate
this, so please stay updated.
```

stevhliu commented on docs/persistence.rst:
```diff
@@ -0,0 +1,80 @@
+code using :func:`eval`. We do, however, have concrete plans on how to mitigate
+this. Please stay updated.
+
+On top of trying to support all of sklearn, we plan on making persistence
```
```suggestion
On top of trying to support persistence in sklearn, we plan on making persistence
```

stevhliu commented on docs/persistence.rst:
```diff
@@ -0,0 +1,80 @@
+work. However, most ``numpy`` and ``scipy`` functions should work. Therefore,
+you can actually save built-in functions like``numpy.sqrt``.
+
+Goals
```
Hmm, I'd maybe rename this section to just Roadmap and combine it with the section below.

stevhliu reviewed (commented):
Good job and super cool docs!

BenjaminBossan commented:
@stevhliu Thanks, your input is really appreciated. I especially tend to use value words, thanks for pointing that out. I addressed all of your points.

adrinjalali commented on docs/persistence.rst:
```diff
@@ -0,0 +1,81 @@
+
+.. warning::
+
+   This feature is very early in development, which means the API is
+   unstable and it is **not secure** at the moment. Therefore, use the same
+   caution as you would for ``pickle``: Don't load from sources that you
+   don't trust. In the future, more security will be added.
```
we should also tag this as experimental, the same way we do in sklearn, so that it's clear from users' code that this is an experimental feature.

BenjaminBossan replied:
I don't understand how that relates to the documentation. Can we maybe work on this in another PR?

adrinjalali replied:
Fair enough. Created this: https://github.com/skops-dev/skops/issues/153

adrinjalali commented on docs/persistence.rst:
```diff
@@ -0,0 +1,81 @@
+load sklearn models without using ``pickle``.
+
+Pickle is the standard serialization format for sklearn and for Python in
+general. One of its biggest advantages is it can be used for almost all Python
```
The sentence is hard for me to parse, maybe:

```suggestion
general. One of the main advantages of ``pickle`` is that it can be used for almost all Python
```

adrinjalali commented on docs/persistence.rst:
```diff
@@ -0,0 +1,81 @@
+code but this flexibility also makes it inherently insecure. This is because
+loading certain types of objects requires the ability to run arbitrary code,
+which can be misused for malicious purposes. For example, an attacker can use it
+to steal secrets from your machine or install a virus. As the Python docs say:
```
link to the python docs?

adrinjalali commented on docs/persistence.rst:
```diff
@@ -0,0 +1,81 @@
+which can be misused for malicious purposes. For example, an attacker can use it
+to steal secrets from your machine or install a virus. As the Python docs say:
+
+    The pickle module is not secure. Only unpickle data you trust. It is
+    possible to construct malicious pickle data which will execute arbitrary
+    code during unpickling. Never unpickle data that could have come from an
+    untrusted source, or that could have been tampered with.
```
maybe put this in a warning tag? (https://sublime-and-sphinx-guide.readthedocs.io/en/latest/notes_warnings.html#warnings)

adrinjalali commented on docs/persistence.rst:
```diff
@@ -0,0 +1,81 @@
+estimator that does not work, please open an issue on the skops `GitHub page
+<https://github.com/skops-dev/skops/issues>`_ and let us know.
+
+In contrast to pickle, skops cannot persist arbitrary Python code. This means
```
We can probably add support while making it explicit that they are loading arbitrary code.

```suggestion
In contrast to pickle, at the moment skops cannot persist arbitrary Python code. This means
```

adrinjalali replied:
there might be a way a way to support certain things by using `inspect.getsource`. I'm still not sure what it'd support, but we can have an explicit `arbitrary_code` type which we always warn on.

BenjaminBossan replied:
How about we change the wording once we're there? :)

adrinjalali replied:
Sure.

adrinjalali commented on skops/io/_persist.py:
```diff
@@ -28,6 +28,29 @@
+
+    Parameters
+    ----------
+    obj
```
```suggestion
    obj: object
```

adrinjalali commented on skops/io/_persist.py:
```diff
@@ -28,6 +28,29 @@
+    Parameters
+    ----------
+    obj
+        The object to be saved
```
```suggestion
        The object to be saved. Usually a scikit-learn compatible model.
```

adrinjalali commented on skops/io/_persist.py:
```diff
@@ -28,6 +28,29 @@
+        The object to be saved
+
+    file: str
+        The file name. A zip archive will automatically created, with the ".zip"
+        ending added to the given filename.
```
we're adding a `skops` extention, not a `zip` one.

BenjaminBossan replied:
Done.

adrinjalali commented on skops/io/_persist.py:
```diff
@@ -43,6 +66,30 @@ def save(obj, file):
+
+    Returns
+    -------
+    instance
```
```suggestion
    instance: object
```

adrinjalali reviewed (commented):
Look good as a first iteration. Thanks @BenjaminBossan

adrinjalali commented on docs/persistence.rst:
```diff
@@ -0,0 +1,84 @@
+almost all Python code but this flexibility also makes it inherently insecure.
+This is because loading certain types of objects requires the ability to run
+arbitrary code, which can be misused for malicious purposes. For example, an
+attacker can use it to steal secrets from your machine or install a virus. As
+the Python docs say:
```
I was thinking something like this would be enough?

```suggestion
attacker can use it to steal secrets from your machine or install a virus. As
the `Python docs <https://docs.python.org/3/library/pickle.html#module-pickle>`__ say:
```

This whole paragraph doesn't have any links to pickle.

BenjaminBossan replied:
This results in the exact same link as ```:mod:`pickle` ``` in the paragraph directly above, but anyway, I added it
